Khan/wonder-blocks
# Fix focus issue with filterable dropdowns

## Summary:
Currently, if a dropdown switches between a virtualized list (125+ results) and
a non-virtualized list (< 125 results), then the search field rerenders and
loses focus. This means that a user is typing in the search field and the filter
results change mid-typing, the focus leaves the text field and the user will have
to click back in to keep typing.

As a temporary fix, here we force focus on the field on mount if the focus was already
previously on the field.

The issue should go away with WB-1310, but we are adding this quick fix as a bandaid for now.

Issue: https://khanacademy.atlassian.net/browse/WB-1281

## Before:
<img width="324" alt="Screen Shot 2022-04-29 at 5 46 49 PM" src="https://user-images.githubusercontent.com/13231763/166083897-2c7cb82f-11ef-4706-a5aa-9491def88e94.png">

https://user-images.githubusercontent.com/13231763/166083939-c26ea043-c6d0-43b0-98b4-10c765965d21.mov


## After:
https://user-images.githubusercontent.com/13231763/166083866-5b3a9284-1444-49e1-8440-8bccfa4f16a7.mov


## Test plan:

Storybook
- Go to storybook
- Go to Dropdown/SingleSelect/With Filter Opened
- Type in "fruit # 5 banana" and confirm that the focus doesn't leave the search field.
- Backspace back to "fruit" and confirm that the focus doesn't leave the search field.

NOTE: I tried adding a unit test in `dropdown-core.test.js` and `single-select.test.js` but it doesn't look like there's a good way to test the behavior with the current implementation of SingleSelect/Dropdown and the current tools in React Testing Library. I wrote some tests that looked like they were intuitively testing the behavior, but when I commented out the bug fix, the tests didn't fail like they were supposed to, due to how the rendering works with dropdowns. So they wouldn't even help prevent this from happening again. While it would be ideal to have unit tests for this, Dropdowns are supposed to be refactored soon as part of WB-1310, so hopefully it's okay to forego adding tests for now.
